src/java/org/apache/cassandra/io/util/ThreadLocalReadAheadBuffer.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.io.util;
+
+import java.nio.ByteBuffer;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import io.netty.util.concurrent.FastThreadLocal;
+import org.apache.cassandra.io.compress.BufferType;
+import org.apache.cassandra.io.compress.CorruptBlockException;
+import org.apache.cassandra.io.sstable.CorruptSSTableException;
+
+public final class ThreadLocalReadAheadBuffer
+{
+    private static class Block {
+        ByteBuffer buffer = null;
+        int index = -1;
+    }
+
+    private final ChannelProxy channel;
+
+    private final BufferType bufferType;
+
+    private static final Map<String, FastThreadLocal<Block>> allBlocks = new ConcurrentHashMap<>();
+
+    private final FastThreadLocal<Block> blockHolder;
+
+    private final int bufferSize;
+
+    public ThreadLocalReadAheadBuffer(ChannelProxy channel, int bufferSize, BufferType bufferType)
+    {
+        this.channel = channel;
+        this.bufferSize = bufferSize;
+        this.bufferType = bufferType;
+        allBlocks.putIfAbsent(channel.filePath(), new FastThreadLocal<>() {
+            @Override
+            public Block initialValue()
+            {
+                return new Block();
+            }
+        });
+        blockHolder = allBlocks.get(channel.filePath());
+    }
+
+    public boolean hasBuffer()
+    {
+        return blockHolder.get().buffer != null;
+    }
+
+    /**
+     * Safe to call only if {@link #hasBuffer()} is true
+     */
+    public int remaining()
+    {
+        return getBlock().buffer.remaining();
+    }
+
+    public void allocateBuffer()
+    {
+        getBlock();
+    }
+
+    private Block getBlock()
+    {
+        Block block = blockHolder.get();
+        if (block.buffer == null)
+        {
+            block.buffer = bufferType.allocate(bufferSize);
+            block.buffer.clear();
+        }
+        return block;
+    }
+
+    public void fill(long position) throws CorruptBlockException
+    {
+        Block block = getBlock();
+        ByteBuffer blockBuffer = block.buffer;
+        long channelSize = channel.size();
```

Comment:
channel.size() causes a system call, assuming that we have a read-only file we can get it once and store as a field of ThreadLocalReadAheadBuffer

<img width="1774" alt="image" src="https://github.com/user-attachments/assets/d73469b4-f3a6-4d14-a8a6-3710bbaa9e88" />

Reply:
see https://github.com/apache/cassandra/pull/3606/commits/38d47aae5d6a2ee36fdf93ba3c467e9902043964